### PR TITLE
[WIP] Dev coios 250 mb way component

### PR DESCRIPTION
**Pull request by mohammedDehairy**
## Summary
1. Created the MBWay component.
2. Created the Await Action.
3. Created the AwaitView, AwaitViewController, and AwaitComponentViewModel.
4. Created a network stack to poll the MBWay api.

## Tested scenarios

**Fixed issue**:

### CODE REVIEW

File: Adyen/Core/APIClient/BackoffScheduler.swift [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+/// :nodoc:
+/// Decides what interval should be used to delay the next event, based on how many times the event has been fired before.
+public protocol IntervalCalculator {
+    func interval(for counter: UInt) -> DispatchTimeInterval
+}
+
+/// :nodoc:
+/// 2 seconds between the first 20 events, and then 10 seconds between the next 80 events,
+/// then `DispatchTimeInterval.never` is returned after that.
+public struct BackoffIntervalCalculator: IntervalCalculator {
```
Comment (descorp) [inline Adyen/Core/APIClient/BackoffScheduler.swift]: Where this rules comes from?
Does anyone actually going to wait payment confirmation for 15 minutes? :D 
maybe better something Fibonacci based?

File: Adyen/Core/APIClient/PaymentStatusRequest.swift [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    internal let path: String = "checkoutshopper/services/PaymentInitiation/v1/status"
+    
+    /// :nodoc:
+    internal var counter: UInt = 0
+    
+    /// :nodoc:
+    internal let headers: [String: String] = ["Origin": "https://test.org", "Content-Type": "application/json"]
+    
+    /// :nodoc:
+    internal let queryParameters: [URLQueryItem] = [URLQueryItem(name: "token", value: "test_YP5RJK3TL5EELNPAS7RHIRVRAIMYZXOH")]
```
Comment (descorp) [inline Adyen/Core/APIClient/PaymentStatusRequest.swift:26]: this should be set from outside, I believe ? :D
Maybe as part of `Environment` object?

File: Adyen/Components/MB Way/MBWayAwaitComponent.swift [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        case let .failure(error):
+            delegate?.didFail(with: error, from: self)
+            return false
+        }
+    }
+    
+    /// :nodoc:
+    private func shouldRetry(_ response: PaymentStatusResponse, paymentData: String) -> Bool {
+        switch response.resultCode {
+        case .pending, .received:
```
Comment (descorp) [inline Adyen/Components/MB Way/MBWayAwaitComponent.swift:65]: what if there is an `PaymentResultCode.error` ?
Should it be `delegate?.didFail(with: error, from: self)` ?

File: Adyen/Helpers/URLSessionHelpers.swift [hunk 4]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2019 Adyen N.V.
+// Copyright (c) 2020 Adyen N.V.
```
Comment (descorp) [inline Adyen/Helpers/URLSessionHelpers.swift:2]: We need to ask IS fro swift-format again 😅

File: Adyen/UI/Await View Controller/AwaitComponentViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,46 @@
+//
+// Copyright (c) 2020 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import Foundation
+
+/// Describes the `AwaitViewController` UI elements.
+internal struct AwaitComponentViewModel {
```
Comment (descorp) [inline Adyen/UI/Await View Controller/AwaitComponentViewModel.swift:10]: I don't think we ever used ViewModel in SDK before...

File: Adyen/Validators/EmailValidator.swift [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        super.init(minimumLength: 6, maximumLength: 320)
+    }
+    
+    /// :nodoc:
+    public override func isValid(_ value: String) -> Bool {
+        guard super.isValid(value) else { return false }
+        // Regex constructed according to the definition of an email address
+        // https://en.wikipedia.org/wiki/Email_address
+        // swiftlint:disable:next line_length
+        let regex = #"^(([^<>\(\)\[\]\\.,;:\s@"]+(\.[^<>\(\)\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$"#
```
Comment (descorp) [inline Adyen/Validators/EmailValidator.swift:23]: I didn't know you can use `\` in email :D

Comment (descorp) [reply Adyen/Validators/EmailValidator.swift:23]: I didn't know you can use `#` for regex :D

File: Adyen/Components/Await/AwaitComponent.swift [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        paymentMethodSpecificAwaitComponent?.handle(action)
+    }
+    
+    /// :nodoc:
+    private func buildPaymentMethodSpecificAwaitComponent(_ paymentMethodType: AwaitPaymentMethod) -> AnyAwaitComponent {
+        switch paymentMethodType {
+        case .mbway:
+            let scheduler = BackoffScheduler(queue: .main)
+            let baseAPIClient = APIClient(environment: environment)
+            let apiClient = self.apiClient ?? RetryAPIClient(apiClient: baseAPIClient, scheduler: scheduler)
```
[the remark this answers is not in the document]
Comment (mohammedDehairy) [reply Adyen/Components/Await/AwaitComponent.swift:83]: actually the `AwaitComponent` with respect to MBWay won't function without a `RetryAPIClient` with a `BackoffScheduler`, the logic to inject an `AnyRetryAPIClient ` is just for unit tests, client code cannot inject `AnyRetryAPIClient`, the initializer that accepts `AnyRetryAPIClient` is internal.

so yes, the `RetryAPIClient` with a `BackoffScheduler` is integral of how the `AwaitComponent` works.

```
I think AwaitComponent should not be aware of environment, specific types of APIClient or how to build one.
```

I disagree :-) .

File: Adyen/Helpers/URLSessionHelpers.swift (hunk 4 above)
Comment (mohammedDehairy) [reply Adyen/Helpers/URLSessionHelpers.swift:2]: @joostvandijk

**Pull request merged**